server/core/static/external/cred_update.js

```diff
@@ -0,0 +1,106 @@
+// Makes the password form interactive (e.g. shows when passwords don't match)
+function setupInteractivePwdFormListeners() {
+    const new_pwd = document.getElementById("new-password");
+    const new_pwd_check = document.getElementById("new-password-check");
+    const pwd_submit = document.getElementById("password-submit");
+
+    function markPwdCheckValid() {
+        new_pwd_check.classList.remove("is-invalid");
+        new_pwd_check.classList.add("is-valid");
+        pwd_submit.disabled = false;
+    }
+
+    function markPwdCheckInvalid() {
+        new_pwd_check.classList.add("is-invalid");
+        new_pwd_check.classList.remove("is-valid");
+        pwd_submit.disabled = true;
+    }
+
+    new_pwd.addEventListener("input", (_) => {
+        // Don't mark invalid if user didn't fill in the confirmation box yet
+        // Also my password manager (keepassxc with autocomplete)
+        //   likes to fire off input events when both inputs were empty.
+        if (new_pwd_check.value !== "") {
+            if (new_pwd.value === new_pwd_check.value) {
+                markPwdCheckValid();
+            } else {
+                markPwdCheckInvalid();
+            }
+        }
+        new_pwd.classList.remove("is-invalid");
+    });
+
+    new_pwd_check.addEventListener("input", (_) => {
+        // No point in updating the status if confirmation box is empty
+        if (new_pwd_check.value === "") return;
+        if (new_pwd_check.value === new_pwd.value) {
+            markPwdCheckValid();
+        } else {
+            markPwdCheckInvalid();
+        }
+    });
+}
+
+function stillSwapFailureResponse(event) {
+    if (event.detail.xhr.status === 422) {
+        console.log("Still swapping failure response")
+        event.detail.shouldSwap = true;
+        event.detail.isError = false;
+    }
+}
+
+function startPasskeyEnrollment() {
+    let data_elem = document.getElementById('data');
```

Comment:
If it's never reassigned, it should be a const
```suggestion
    const data_elem = document.getElementById('data');
```

Reply:
Also we're assuming that the getElementById and then parse work, which probably isn't great - what happens if it's not there for some reason?

Reply:
JS logs an exception and clicking the button will seemingly do nothing.
The element with this id sits in the same file that calls passkeyNamingSubmitBtn: add_passkey_partial.html

I can show an alert browser popup telling them we have some internal data structure mismatch and that they should report this to us ?

Reply:
I'd check for the fail to get by id and parsing errors and log that to the console, because the exception can cause all javascript to stop, which isn't great.

Then pop a JS confirm saying "There was an error handling a server response, would you like to try and reload the page?" which when they click OK, reloads the tab - which should re-initialise everything.